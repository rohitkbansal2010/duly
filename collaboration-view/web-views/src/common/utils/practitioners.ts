import uniqBy from 'lodash/uniqBy';

import {
  MAX_PETAL_PRACTITIONERS_COUNT,
  SEPARATED_PETAL_PRACTITIONERS_COUNT
} from '@constants';
import {
  CommonPractitionerType,
  ExtendedCommonPractitionerType,
  AppointmentData,
  Practitioner
} from '@types';

import { mergeHumanName } from './person';

const getIsNeedToSeparate = (practitioners: CommonPractitionerType[]): boolean =>
  practitioners.length > MAX_PETAL_PRACTITIONERS_COUNT;

const getFirstPractitioners = (
  practitioners: CommonPractitionerType[]
): CommonPractitionerType[] =>
  getIsNeedToSeparate(practitioners)
    ? practitioners.slice(0, SEPARATED_PETAL_PRACTITIONERS_COUNT)
    : practitioners;

const getOtherPractitioners = (
  practitioners: CommonPractitionerType[]
): CommonPractitionerType[] =>
  getIsNeedToSeparate(practitioners)
    ? practitioners.slice(SEPARATED_PETAL_PRACTITIONERS_COUNT, practitioners.length)
    : [];

export const getPractitionersGroups = (
  practitioners: ExtendedCommonPractitionerType[]
): { [key: string]: ExtendedCommonPractitionerType[] } =>
  ({
    firstPractitioners: getFirstPractitioners(practitioners),
    otherPractitioners: getOtherPractitioners(practitioners),
  });

export const getPractitionersFromAppointments = (
  appointments: AppointmentData[]
): Practitioner[] => {
  const allPractitioners = appointments.map(({ practitioner }) =>
    practitioner);
  const uniqPractitioners = uniqBy(allPractitioners, 'id');

  return uniqPractitioners;
};

export const getPractitionersWithoutUser = (
  practitioners: Practitioner[],
  user: Practitioner
): Practitioner[] =>
  practitioners.filter(({ humanName }) =>
    mergeHumanName(humanName) !== mergeHumanName(user.humanName));

export const setActiveUser = (
  careTeamMembers: CommonPractitionerType[]
): ExtendedCommonPractitionerType[] =>
  careTeamMembers.map((member, index) =>
    ({ ...member, isCurrentUser: index === 0 }));

export const getPractitionersSuffix = (suffixes: string[]): string => {
  let specialty = '';
  suffixes.forEach((suffix) => {
    specialty += suffix + ', ';
  });
  
  specialty = specialty.slice(0, specialty.length - 2);
  return specialty;
};
