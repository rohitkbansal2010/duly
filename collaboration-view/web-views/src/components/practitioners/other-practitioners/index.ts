export { OtherPractitioners } from './other-practitioners';
