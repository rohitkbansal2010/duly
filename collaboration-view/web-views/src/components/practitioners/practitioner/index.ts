export { Practitioner } from './practitioner';
export type { } from './practitioner';
