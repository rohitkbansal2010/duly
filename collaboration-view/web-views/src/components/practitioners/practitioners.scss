.practitioners {
  display: flex;
  align-items: center;
}

.flexRow {
  flex-direction: row;
}

.flexColumn {
  flex-direction: column;
}
