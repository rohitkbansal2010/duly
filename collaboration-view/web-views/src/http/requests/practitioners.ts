import { HttpMethod } from '@enums';
import { EncouterServices, PatientsService } from '@http/api';
import { Practitioner } from '@types';
import { httpRequest } from '@utils';

export const getBrief = (patientId: number): Promise<Practitioner | undefined> =>
  httpRequest({
    url: `${EncouterServices.PRACTITIONERS}/${patientId}${PatientsService.BRIEF}`,
    method: HttpMethod.GET,
  });
