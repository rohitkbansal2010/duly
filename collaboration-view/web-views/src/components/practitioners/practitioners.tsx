import { ExtendedCommonPractitionerType } from '@types';
import { getPractitionersGroups } from '@utils';

import { OtherPractitioners } from './other-practitioners';
import { Practitioner } from './practitioner';
import styles from './practitioners.scss';

type PractitionersPropsType = {
  practitioners: ExtendedCommonPractitionerType[];
  isTodaysAppointment?: boolean;
};

export const Practitioners = ({
  practitioners,
  isTodaysAppointment = false,
}: PractitionersPropsType) => {
  const { firstPractitioners, otherPractitioners } = getPractitionersGroups(practitioners);

  const renderPractitionersList = ({
    id,
    isCurrentUser,
    ...props
  }: ExtendedCommonPractitionerType) =>
    (
      <Practitioner
        key={id}
        id={id}
        role={'role' in props ? props.role : props.memberType}
        {...props}
        isCurrentUser={isCurrentUser}
        isTodaysAppointment={isTodaysAppointment}
      />
    );

  const practitionersFlex = isTodaysAppointment ?
    styles.flexColumn :
    styles.flexRow;

  return (
    <div
      className={`${styles.practitioners} ${practitionersFlex}`}
    >
      {firstPractitioners.map(renderPractitionersList)}
      {!!otherPractitioners.length && (
        <OtherPractitioners
          practitioners={otherPractitioners}
          isTodaysAppointment={isTodaysAppointment}
        />
      )}
    </div>
  );
};
