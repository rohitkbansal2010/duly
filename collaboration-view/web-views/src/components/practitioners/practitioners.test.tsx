import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { PractitionerRole } from '@enums';
import { Practitioner } from '@types';

import { Practitioners } from './practitioners';

const mockPractinioners: Practitioner[] = [
  {
    id: 'qwerty1',
    humanName: {
      familyName: 'Morris',
      givenNames: [ 'Jane' ],
    },
    photo: {
      contentType: 'image/x-png',
      title: 'Photo',
      size: 0,
      data: 'data',
    },
    role: PractitionerRole.MA,
  },
  {
    id: 'qwerty2',
    humanName: {
      familyName: 'Ling',
      givenNames: [ 'Jane' ],
      prefixes: [ 'Dr.' ],
    },
    photo: {
      contentType: 'image/x-png',
      title: 'Photo',
      size: 0,
      data: 'data',
    },
    role: PractitionerRole.PCP,
  },
];

describe('Practitioners', () => {
  let component: ReturnType<typeof shallow>;

  beforeEach(() => {
    component = shallow(
      <Practitioners
        practitioners={mockPractinioners}
      />
    );
  });

  it('should exist', () => {
    expect(Practitioners).toBeDefined();
  });

  it('should render practitioners', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should contains one practitioner with role MA', () => {
    const practitionerWithRoleMA = mockPractinioners.filter(({ role }) =>
      role === PractitionerRole.MA);

    expect(
      component.children().findWhere(child =>
        child.props().role === PractitionerRole.MA)
    ).toHaveLength(practitionerWithRoleMA.length);
  });

  it('should contains practitioners with role PCP', () => {
    const practitionerWithRolePCP = mockPractinioners.filter(({ role }) =>
      role === PractitionerRole.PCP);

    expect(
      component.children().findWhere(child =>
        child.props().role === PractitionerRole.PCP)
    ).toHaveLength(practitionerWithRolePCP.length);
  });
});
