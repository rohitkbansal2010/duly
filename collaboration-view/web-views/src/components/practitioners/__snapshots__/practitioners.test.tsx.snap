// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Practitioners should render practitioners 1`] = `
<div
  className="practitioners flexRow"
>
  <Practitioner
    humanName={
      Object {
        "familyName": "Morris",
        "givenNames": Array [
          "Jane",
        ],
      }
    }
    id="qwerty1"
    isTodaysAppointment={false}
    key="qwerty1"
    photo={
      Object {
        "contentType": "image/x-png",
        "data": "data",
        "size": 0,
        "title": "Photo",
      }
    }
    role="MA"
  />
  <Practitioner
    humanName={
      Object {
        "familyName": "Ling",
        "givenNames": Array [
          "Jane",
        ],
        "prefixes": Array [
          "Dr.",
        ],
      }
    }
    id="qwerty2"
    isTodaysAppointment={false}
    key="qwerty2"
    photo={
      Object {
        "contentType": "image/x-png",
        "data": "data",
        "size": 0,
        "title": "Photo",
      }
    }
    role="PCP"
  />
</div>
`;
